### PR TITLE
add localstorage-writable

### PR DESCRIPTION
**Pull request by jjnp**
Adds a fully compatible custom store that persists all values of the writable to local storage.
Caveat is of course that the value has to be serializable with JSON. An alternative might be to use devalue?

### CODE REVIEW

File: src/lib/utils/localstorage-writable.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+		}
+
+		set(value)
+	}
+
+	const wipe = () => {
+		if (browser) {
+			set(null)
+			localStorage.removeItem(localStorageKey)
+		}
```
Comment (benjaminstrasser) [inline src/lib/utils/localstorage-writable.ts:37]:
```suggestion
		if (browser) {
			localStorage.removeItem(localStorageKey)
		}
		set(null)
```
More consistency imo to setAndStore where localStorage is ignored but store is updated.

**Review by benjaminstrasser — CHANGES_REQUESTED** [review]

File: src/lib/utils/localstorage-writable.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ * @param localStorageKey The key used for local-storage
+ * @param defaultValue The default value in case there is none found in localstorage
+ * @returns A standard writable that persists all value changes to localstorage
+ */
+export const localStorageWritable = <T>(
+	localStorageKey: string,
+	defaultValue?: T
+): LocalStorageWritable<T> => {
+	const storedValue = browser ? localStorage.getItem(localStorageKey) : null
+	const localStorageValue = storedValue !== null ? JSON.parse(storedValue) : null
```
[the remark this answers is not in the document]
Comment (jjnp) [reply src/lib/utils/localstorage-writable.ts:19]: No you're right, I prefer the concise version. Changed it.

**Review by benjaminstrasser — COMMENTED** [review]
I wrote some unit tests, which also identified an issue, where `localStorageWritable(key, defaultValue)` adds the value to the store but does not insert the default value into localstorage.

```typescript
import { localStorageWritable } from './localstorage-writable'
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest'
import { get } from 'svelte/store'

const mockedBrowser = vi.hoisted(() => {
	return vi.fn()
})

vi.mock('$app/environment', () => {
	return {
		get browser() {
			return mockedBrowser()
		}
	}
})

const localStorageMockFactory = () => {
	let store: Record<string, string> = {}
	return {
		getItem: (key: string) => store[key] || null,
		setItem: (key: string, value: string) => {
			store[key] = value
		},
		removeItem: (key: string) => {
			delete store[key]
		}
	}
}

describe('localStorageWritable', () => {
	beforeEach(() => {
		mockedBrowser.mockReturnValue(true)
		vi.stubGlobal('localStorage', localStorageMockFactory())
	})

	afterEach(() => {
		vi.resetAllMocks()
	})

	it('should initialize with default value if no value in localStorage', () => {
		const defaultValue = 'default'

		const store = localStorageWritable('testKey', defaultValue)

		expect(get(store)).toBe(defaultValue)
		expect(localStorage.getItem('testKey')).toBe(JSON.stringify(defaultValue))
	})

	it('should initialize with value from localStorage if present', () => {
		localStorage.setItem('testKey', JSON.stringify('storedValue'))

		const store = localStorageWritable('testKey')

		expect(get(store)).toBe('storedValue')
	})

	it('should prioritize value from local storage over default', () => {
		localStorage.setItem('testKey', JSON.stringify('storedValue'))

		const store = localStorageWritable('testKey', 'defaultValue')

		expect(get(store)).toBe('storedValue')
		expect(localStorage.getItem('testKey')).toBe(JSON.stringify('storedValue'))
	})

	it('should store value in localStorage when set', () => {
		const store = localStorageWritable('testKey')
		store.set('newValue')

		expect(localStorage.getItem('testKey')).toBe(JSON.stringify('newValue'))
		expect(get(store)).toBe('newValue')
	})

	it('should remove value from localStorage when wipe is called', () => {
		localStorage.setItem('testKey', JSON.stringify('storedValue'))

		const store = localStorageWritable('testKey')
		store.wipe()

		expect(localStorage.getItem('testKey')).toBe(null)
		expect(get(store)).toBe(null)
	})

	it('should update the value correctly using update method', () => {
		const store = localStorageWritable<number>('testKey', 1)
		store.update((n) => n! + 1)

		expect(localStorage.getItem('testKey')).toBe(JSON.stringify(2))
		expect(get(store)).toBe(2)
	})
})

describe('localStorageWritable with no browser', () => {
	beforeEach(() => {
		mockedBrowser.mockReturnValue(false)
		vi.stubGlobal('localStorage', undefined)
	})

	afterEach(() => {
		vi.resetAllMocks()
	})

	it('should initialize with default value', () => {
		const defaultValue = 'default'
		const store = localStorageWritable('testKey', defaultValue)

		expect(get(store)).toBe(defaultValue)
	})

	it('should initialize with null', () => {
		const store = localStorageWritable('testKey')

		expect(get(store)).toBe(null)
	})

	it('should set value', () => {
		const store = localStorageWritable('testKey')
		store.set('newValue')

		expect(get(store)).toBe('newValue')
	})

	it('should remove value', () => {
		const store = localStorageWritable('testKey')
		store.set('newValue')
		store.wipe()

		expect(get(store)).toBe(null)
	})

	it('should update the value correctly using update method', () => {
		const store = localStorageWritable<number>('testKey', 1)
		store.update((n) => n! + 1)

		expect(get(store)).toBe(2)
	})
})

```

**Comment by sjaghori** [comment]
For what kind of use cases are we going to utilize this?

**Comment by benjaminstrasser** [comment]
> For what kind of use cases are we going to utilize this?

I think it was intended for #77

**Comment by jjnp** [comment]
@benjaminstrasser  Pls directly push the unit-tests, I'll look at the issue (but ofc feel free to add a fix directly, if you have one)

**Comment by jjnp** [comment]
> For what kind of use cases are we going to utilize this?

Right, it's intended for the other PR, but it comes in handy pretty often. Might be some things and local configs like dark mode settings, preferences, undo history, recent search keywords, ...

**Review by benjaminstrasser — APPROVED** [review]

**Review by mledl — APPROVED** [review]
Nice utility and good testing 🚀

File: src/lib/utils/localstorage-writable.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+ * @returns A standard writable that persists all value changes to localstorage
+ */
+export const localStorageWritable = <T>(
+	localStorageKey: string,
+	defaultValue?: T
+): LocalStorageWritable<T> => {
+	const storedValue = browser ? localStorage.getItem(localStorageKey) : null
+	const localStorageValue = (() => {
+		try {
+			return storedValue !== null ? JSON.parse(storedValue) : null
```
Comment (sjaghori) [inline src/lib/utils/localstorage-writable.ts:21]: @benjaminstrasser the check here is redundant. 
`JSON.parse(null) === null`

Comment (benjaminstrasser) [reply src/lib/utils/localstorage-writable.ts:21]: Yes you are correct! I caught that too. Problem is JSON.parse is typed as `parse(value: string)` and not `parse(value: string | null)` even though `parse(null) === null`.

See here: https://github.com/codingcommons/tiny-tms/pull/81#discussion_r1623911640

**Pull request merged**